### ISSUE TITLE
documentation: need to document sharded topology

### ISSUE THREAD

**Issue by krancour** [akuity/kargo#2447]

**Comment by krancour** [comment]
Yes. It's fully possible and has been for a very long time. The issue is indeed only that it isn't documented. I know there are plenty of people in the community eager for this documentation, but internally it hasn't managed to eclipse other (customer-driven) priorities. And while we're grateful to anyone in the community who's expressed willingness to work on this (you're not the first), we think this is a complicated enough topic that we'd prefer it be handled directly by a maintainer.

If you're just looking to set this up for yourself, this should help get you going:

- The controller (_not_ the management controller, but the "regular" one) is the only component that needs to be distributed around to different clusters.

- Each distributed controller basically needs to talk to _two_ clusters -- one being the cluster hosting the Kargo control plane and the other being the cluster hosting an Argo CD control plane. Typically the distributed controllers would run directly in those clusters hosting Argo CD control planes. This "phone home to the Kargo control plane" approach means the Kargo control plane doesn't have highly-privileged credentials for all your other clusters and does not become a high-value target.

- You can use the normal chart to install the control plane and _also_ to install the distributed controllers. For the latter, you leave only that one component enabled and disable all others. The chart's `values.yaml` has a lot of instructive comments on how configure things.

**Comment by brandonjbjelland** [comment]
Thank you! This is excellent guidance and confers much of what I've discovered on my own.

**Comment by ilyakazm** [comment]
@krancour, we are migrating our CI/CD system to ArgoCD/Kargo and want to use sharded/distributed controllers in the following scenario:

* Kargo Control Plane runs on "primary" cluster (with all the components - API, Management Controller etc), next to ArgoCD
* Distributed controllers run on remote clusters where all the actual organization's applications/code are run
* In Production environment, each application deploys Kargo CRs (Project, WHs, QA/Stg/Prod Stages etc) via company's maintained Helm chart's templates  _inside each application's namespace_ (Namespaces would be [labeled](https://docs.kargo.io/user-guide/how-to-guides/working-with-projects#namespace-adoption) with Kargo's label to adopt them as we need it as per organizational requirements)
* Distributed Kargo Controller discovers the CRs and "phone home to the Kargo control plane" making them visible in Kargo's UI
* Devs see their Projects in Kargo UI and do the promotions etc

Our in-house Helm chart is heavily customized and used as a base for 99% of our applications with all the necessary templated CRs included, so we thought to include Kargo's CRs as templates and manage them like that. I understand that this might be not the optimal/desired approach, especially for Kargo, but the goal here is to eliminate the need for devs to go to other (mono)repos to add kargo configs but to have all surrounding configuration for their apps in one place (app's repo) if possible. 

Would the above scenario work with current Kargo's functionality? 

I've tried the suggested [setup](https://github.com/akuity/kargo/issues/2447#issuecomment-2860636652) above (combined with [these configs](https://github.com/akuity/kargo/issues/3660#issue-2925201645) with kubeconfigs provided etc and it seems that if I create, say, Project on a remote cluster where distributed controller is running, in some test application's namespace, controller tries to create a namespace on _"primary"_ cluster instead. Either I do something wrong or maybe that's by design - not sure.

Apologies if anything is unclear - happy to provide more details if needed! 

Thanks in advance!

**Comment by gnadaban** [comment]
When this is documented it would be great to also include instructions and topology for eg. Argo Rollouts.
I've been trying to make verifications work in a multi-cluster topology expecting that analysisruns are created in the cluster where the sharded controller runs, without luck so far. Based on the [code that starts verifications](https://github.com/akuity/kargo/blob/fc3047e6f48406f027414483f01bf8c9f34578a1/internal/controller/stages/regular_stages.go#L1373C1-L1383C3) I have an inkling that it's currently only supporting creating runs in the control plane cluster.

Not sure what the intent or design is, so I might be wrong, so clarification here would be great. :)

**Comment by joebowbeer** [comment]
@krancour Is this the issue that will be fixed when the (empty) [Architecture](https://docs.kargo.io/operator-guide/architecture) page linked in the FAQ is fleshed-out?

> ### How do I integrate with multiple Argo CD control planes?
> 
> To get an overview of how this can be achieved, head on over to our [Architecture](https://docs.kargo.io/operator-guide/architecture) section to learn about the topology of a large-scale Kargo deployment.

**Comment by krancour** [comment]
@joebowbeer indeed it is. And to be candid, I feel quite badly that it's something I _still_ haven't gotten around to.

As one might gather from reading through myriad issues that mention this, there be dragons here.

* Documenting this accurately requires some really deep Kargo knowledge. The limited number of people who can speak to this topic authoritatively have been perpetually preoccupied with higher priorities.

* We (Akuity) only wish to document it _to a certain minimum extent_. Personally, I'd like to be transparent about _how_ it works (high-level explanation, diagrams, etc.), but want to stop short of exhaustively documenting the admittedly onerous process of setting this up because:

  * We want to manage expectations. We're not staffed to provide free support to the community on so complex a topic.
 
  * The fact of Akuity Platform seamlessly managing what is otherwise a rather onerous configuration effort is a bit of a selling point for our paid product and we don't want to undercut ourselves too deeply.

Our 1.7 development cycle is shaping up to be heavily oriented around paying down some debts, so hopefully this is something I can squeeze into there. 🤞
